### PR TITLE
Initial port of patrick's branch for CPP

### PR DESCRIPTION
**Pull request by coltfred**
TODO:

- [x] Tests

### CODE REVIEW

File: common/build.rs [hunk 1]
```diff
@@ -91,42 +98,75 @@ fn gen_binding<P: AsRef<Path>>(
...
 fn expand_equals_and_hashcode_macro(out: &str) {
-    let equals_and_hashcode = r##"method hash(&self) -> i32; alias hashCode;
-    private method eq(&self, o: &$1) -> bool; alias rustEq;
-    foreign_code r#"
-    public boolean equals(Object obj) {
-        if(obj instanceof $1){
-            $1 other = ($1) obj;
-            return other.rustEq(this);
+    cfg_if::cfg_if! {
+        if #[cfg(cpp)]{
```
Comment (giarc3) [inline common/build.rs:154]: this won't work as expected
```suggestion
        if #[cfg(feature = "cpp")]{
```

File: cpp/chrono-include.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+mod swig_foreign_types_map {}
```
Comment (giarc3) [inline cpp/chrono-include.rs:1]: Our other `chrono-include.rs` currently lives in `common`. Do you think we should use compilation flags and combine them? Or move the java one out of common and put it in each workspace member?

File: common/build.rs [hunk 3]
```diff
@@ -91,42 +98,75 @@ fn gen_binding<P: AsRef<Path>>(
...
+                "com.ironcorelabs.sdk".into(),
+            )))
+            .merge_type_map("chrono_support", include_str!("chrono-include.rs"));
+        }
+    }
+
+    swig_gen
+        .rustfmt_bindings(true)
+        .remove_not_generated_files_from_output_directory(true) //remove outdated *.java or cpp files
+        .expand("rust_swig_test_jni", from, out_dir.join("lib.rs"));
```
[the remark this answers is not in the document]
Comment (coltfred) [reply common/build.rs]: Well, it never got changed from `test_jni` It looks like it's used for reporting errors if there are any?

https://docs.rs/rust_swig/0.4.0/src/rust_swig/lib.rs.html#421

I've never found those errors very useful, but we can set this as a name in the if/else so I'll do that.

File: common/build.rs [hunk 4]
```diff
@@ -91,42 +98,75 @@ fn gen_binding<P: AsRef<Path>>(
...
-        .join("com")
-        .join("ironcorelabs")
-        .join("sdk");
-    if !path.exists() {
-        std::fs::create_dir_all(&path).expect(
-            "Couldn't create codegen output directory at OUT_DIR/java/com/ironcorelabs/sdk.",
-        );
+cfg_if::cfg_if! {
+    if #[cfg(feature = "cpp")]{
+    fn get_cpp_codegen_output_directory() -> PathBuf {
```
[the remark this answers is not in the document]
Comment (coltfred) [reply common/build.rs]: fixed.

File: common/build.rs [hunk 5]
```diff
@@ -91,42 +98,75 @@ fn gen_binding<P: AsRef<Path>>(
...
+        //let out_dir = env::var("OUT_DIR").expect("no OUT_DIR, but cargo should provide it");
+        let path = Path::new("generated").join("sdk");
```
[the remark this answers is not in the document]
Comment (coltfred) [reply common/build.rs]: I don't have any idea (or plan) on how these header files should be packaged. I'll write a ticket to figure that out.

File: common/build.rs [hunk 6]
```diff
@@ -91,42 +98,75 @@ fn gen_binding<P: AsRef<Path>>(
...
+        if !path.exists() {
+          std::fs::create_dir_all(&path).expect("Couldn't create codegen output directory at cpp/sdk/.");
```
[the remark this answers is not in the document]
Comment (coltfred) [reply common/build.rs]: I'll fix that.

File: common/build.rs (hunk 1 above)
Comment (coltfred) [reply common/build.rs]: Fixed.

File: cpp/chrono-include.rs (hunk 2 above)
Comment (coltfred) [reply cpp/chrono-include.rs:1]: It's shared there. I was assuming that common was a place for things shared between the 2 java projects. I guess we could do a cfg_if in there. Do you think that's better?

Comment (giarc3) [reply cpp/chrono-include.rs:1]: I was thinking that `common` would be for anything we could share between members, not just between java members. I think that we should probably share `chrono-include.rs` with cpp as well, just so it's easier to know where to go looking for it

**Comment by clintfred** [comment]
@coltfred I applied some very minor formatting to a few places. I was able to run the `android-cpp` cargo build.

File: cpp/chrono-include.rs (hunk 2 above)
Comment (coltfred) [reply cpp/chrono-include.rs:1]: Hmmm, it didn't seem to work to just do a cfg_if block in that file. If you want to see what I did I committed what I thought should work and then reverted. https://github.com/IronCoreLabs/ironoxide-java/pull/109/commits/b665dce281a6f0db14e8e531de0bbd1e31e8ec66

Comment (giarc3) [reply cpp/chrono-include.rs:1]: Checked it out and played with it. It seems that `foreign_typemap!(...)` works, but `{ foreign_typemap(...) }` fails. So I'd guess that the `cfg_if` is putting in some brackets it doesn't like?
Unfortunately, I couldn't make the conditional compilation work without being able to chunk it with brackets. So if you're happy where it is, I'm fine with that

File: common/build.rs [hunk 7]
```diff
@@ -91,42 +99,81 @@ fn gen_binding<P: AsRef<Path>>(
...
-        .join("ironcorelabs")
-        .join("sdk");
-    if !path.exists() {
-        std::fs::create_dir_all(&path).expect(
-            "Couldn't create codegen output directory at OUT_DIR/java/com/ironcorelabs/sdk.",
-        );
+cfg_if::cfg_if! {
+    if #[cfg(feature = "cpp")]{
+        fn get_cpp_codegen_output_directory() -> PathBuf {
+            //let out_dir = env::var("OUT_DIR").expect("no OUT_DIR, but cargo should provide it");
```
Comment (clintfred) [inline common/build.rs:133]: comment to clean up

File: common/build.rs [hunk 8]
```diff
@@ -91,42 +99,81 @@ fn gen_binding<P: AsRef<Path>>(
...
+            let path = Path::new("generated").join("sdk");
+            if !path.exists() {
+                std::fs::create_dir_all(&path).expect(format!("Couldn't create codegen output directory at {:?}.", path).as_str());
```
Comment (clintfred) [inline common/build.rs:136]: does this get removed by cargo clean? if not, should it?

File: cpp/BUILDING.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
+# C++ Binding instructions
+
+The building of the CPP library is currently not done by CI, so if you're wanting to produce C/C++ bindings and the associated
+library you'll need to follow the following steps.
+
+## Prerequisites
+
+- C++ 17 or greater installed. Most modern gcc or clang compilers should do
+- C++ Boost installed. Search up directions for your distribution if you're not sure.
```
Comment (clintfred) [inline cpp/BUILDING.md:9]: Do we know what pieces of boost we require? On unbuntu installed a meta-package for `libboost-all-dev`, which installed 100+ packages and took 500 MB of space. :-| If we don't know which pieces we need, we should write a ticket to find out.

File: cpp/cmake/rust_part_lib.cmake [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  set(RUST_PART_LIB_PATH "${TARGET_PATH}/release/${RUST_PART_LIB_NAME}")
+else()
+  set(RUST_PART_LIB_PATH "${TARGET_PATH}/debug/${RUST_PART_LIB_NAME}")
+endif()
+
+add_custom_command(OUTPUT ${RUST_PART_LIB_PATH}
+  COMMAND ${CARGO_BUILD}
+  WORKING_DIRECTORY "${RUST_BUILD_CWD}")
+add_custom_target(rust_part_lib_target DEPENDS ${RUST_PART_LIB_PATH})
+if (MSVC)
```
Comment (clintfred) [inline cpp/cmake/rust_part_lib.cmake:90]: I assume we didn't test this on MSVC?

File: cpp/test/test.cpp [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+template <class T>
+T unwrap(std::variant<T, RustString> value)
+{
+    if (value.index() == 0)
+    {
+        return std::get<T>(std::move(value));
+    }
+    else
+    {
+        std::cout << "Got fatal error: " << std::get<RustString>(value).to_string_view() << "\n";
```
Comment (clintfred) [inline cpp/test/test.cpp:24]: This is our equivalent of panic, I guess?

File: cpp/test/test.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    auto rust_error_message = std::get<1>(value);
+    auto error_message = rust_error_message.to_std_string();
+    TEST_CHECK(error_message.length() > 10);
+    TEST_MSG("Error was: %s", error_message.c_str());
+}
+
+void encrypt_decrypt_roundtrip(void)
+{
+    auto s = "{\"accountId\":\"abcABC012_.$#|@/:;=+'-91e078f0-a60c-4251-8652-dd498c07a8f4\",\"segmentId\":1825,\"signingPrivateKey\":\"uKHa70uwLVG3IU7XodT2kla/PuC/En8PkRCjMMc9ZE7HFrOV+g0vOwATp/CiXp65mVas0K6TSl/RaxDGlcmsnA==\",\"devicePrivateKey\":\"YZRlDSkM+JxxSXCtWCVK693qfhNqcbhaPrtHs92uD4w=\"}";
+    DeviceContext d = unwrap(DeviceContext::fromJsonString(s));
```
Comment (clintfred) [inline cpp/test/test.cpp:67]: My C++ is pretty rusty, but do we understand how memory allocation works with these C++ types? Are these all stack allocated and thus don't require any cleanup?

File: cpp/test/test.cpp [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    IronOxide sdk = unwrap(IronOxide::initialize(d, IronOxideConfig()));
+    auto foo = unwrap(sdk.documentEncrypt(string_to_slice("foo"), DocumentEncryptOpts()));
+    unwrap(sdk.documentDecrypt(vec_to_slice(foo.getEncryptedData())));
```
Comment (clintfred) [inline cpp/test/test.cpp:70]: Did you mean to assert about the decrypted value here?

File: cpp/BUILDING.md (hunk 9 above)
Comment (coltfred) [reply cpp/BUILDING.md:9]: We only need optional, but I don't know what that requires.

File: cpp/cmake/rust_part_lib.cmake (hunk 10 above)
Comment (coltfred) [reply cpp/cmake/rust_part_lib.cmake:90]: I didn't. When we do CI we could.

File: cpp/test/test.cpp (hunk 11 above)
Comment (coltfred) [reply cpp/test/test.cpp:24]: It's a hack in our tests for now. We probably shouldn't exit, but when we beef up the tests we can fix that.

File: cpp/test/test.cpp (hunk 12 above)
Comment (coltfred) [reply cpp/test/test.cpp:67]: Certainly not stack allocated. Strings themselves are heap allocations. Are you asking if I should be calling free?

File: cpp/test/test.cpp (hunk 13 above)
Comment (coltfred) [reply cpp/test/test.cpp:70]: Yes, I'll fix that.

File: common/build.rs (hunk 8 above)
Comment (coltfred) [reply common/build.rs:136]: It does not, because it's not under target. I guess we should decide if we want the headers to be under target or not.

**Comment by coltfred** [comment]
I'm merging this so I can start on the follow-up tickets to further beef up the tests and get CI going.

File: cpp/test/test.cpp (hunk 12 above)
Comment (clintfred) [reply cpp/test/test.cpp:68]: That is what I'm asking...

**Pull request merged**